### PR TITLE
Throw a validation error if there are duplicate operationId in the spec

### LINKED ISSUE #85: Validator does not check uniqueness of operation ids
According to the Swagger spec, the `operationId` of an operation object is:

> Unique string used to identify the operation. The id MUST be unique among all operations described in the API. Tools and libraries MAY use the operationId to uniquely identify an operation, therefore, it is recommended to follow common programming naming conventions.

The validator does not currently check that `operationId`s are unique across the API.  This would be a helpful feature because some codegen tools fail if this constraint is not met.

### PR DESCRIPTION
**Pull request by dan98765**
Would resolve #85 
We have something like this internally already; adding it to swagger_spec_valiator. 

Internally we also assert that there is an operationId defined for each api (endpoint), but https://github.com/OAI/OpenAPI-Specification/issues/1019 makes me think this is controversial so I didn't add that part to this PR; if there's no operationId defined just move on.

### CODE REVIEW

**Comment by watterso** [comment]
🥇

**Comment by dan98765** [comment]
Updated such that uniqueness is only enforced per-tag. Please check out the added unit test that clarifies the behavior.

File: swagger_spec_validator/validator20.py [hunk 1]
```diff
@@ -196,6 +197,9 @@ def validate_apis(apis, deref):
     :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
     :raises: :py:class:`jsonschema.exceptions.ValidationError`
     """
+    operation_tag_to_operation_id_set = defaultdict(set)
+    # operation_ids = set()
```
Comment (watterso) [inline swagger_spec_validator/validator20.py]: woopsies commented code, please delete

File: swagger_spec_validator/validator20.py [hunk 2]
```diff
@@ -206,6 +210,22 @@ def validate_apis(apis, deref):
             if oper_name == 'parameters' or oper_name.startswith('x-'):
                 continue
             oper_body = deref(api_body[oper_name])
+            oper_tags = deref(oper_body.get('tags', [None]))
+
+            # Check that, if this operation has an operationId defined,
+            # no other operation with a same tag also has that
+            # operationId.
+            # import ipdb; ipdb.set_trace()
```
Comment (watterso) [inline swagger_spec_validator/validator20.py]: trailing ipdb, delete this too please :)

File: swagger_spec_validator/validator20.py [hunk 3]
```diff
@@ -206,6 +210,22 @@ def validate_apis(apis, deref):
...
+            operation_id = oper_body.get('operationId')
+            if operation_id is not None:
+                for oper_tag in oper_tags:
+                    if operation_id in operation_tag_to_operation_id_set[oper_tag]:
+                        raise SwaggerValidationError(
+                            "Duplicate operationId: {}".format(operation_id)
+                        )
+                    else:
+                        operation_tag_to_operation_id_set[oper_tag].add(operation_id)
```
Comment (watterso) [inline swagger_spec_validator/validator20.py]: style nit: remove the else and un-indent the add line

File: tests/validator20/validate_apis_test.py [hunk 4]
```diff
@@ -152,3 +152,82 @@ def test_api_check_default_fails(partial_parameter_spec, validator, instance):
     validation_error = excinfo.value.args[1]
     assert validation_error.instance == instance
     assert validation_error.validator == validator
+
```
Comment (watterso) [inline tests/validator20/validate_apis_test.py]: unrelated to line - looks like coverage failed?

File: swagger_spec_validator/validator20.py (hunk 1 above)
Comment (dan98765) [reply swagger_spec_validator/validator20.py]: Oh man you are so right I totally got distracted and forgot to check for silly stuff like this. My bad; I'll clean 'er up.

File: swagger_spec_validator/validator20.py (hunk 3 above)
Comment (dan98765) [reply swagger_spec_validator/validator20.py]: donezo

File: tests/validator20/validate_apis_test.py (hunk 4 above)
Comment (watterso) [reply tests/validator20/validate_apis_test.py]: not sure what i was looking at when i posted this bit, gonna figure out how to drop

Comment (watterso) [reply tests/validator20/validate_apis_test.py]: O i was looking at the [travis CI page](https://travis-ci.org/Yelp/swagger_spec_validator/builds/386916806?utm_source=github_status&utm_medium=notification) looks like the latest commit is passing

Comment (watterso) [reply tests/validator20/validate_apis_test.py]: Yikes i don't know how to github, i shouldn't be the one approving the change, pls no merge

**Comment by monospacesoftware** [comment]
Hi! Our builds started failing today with "swagger_spec_validator.common.SwaggerValidationError: Duplicate operationId: get"... 

We use "x-swagger-router-controller" to specify the target class, and "operationId" to specify the function to call on that class. Multiple classes have a "get" function.

```
    "/api/v2/entities/version": {
      "get": {
        "description": "Query API version",
        "operationId": "get",
        "x-swagger-router-controller": "resources.version.Version",
        "produces": [
          "application/json"
        ], ...
```

I'm not going to assert this is the best pattern, but I doubt we'll be the only ones effected by this change.

Using x-swagger-router-controller lets us avoid using the full function path as the operationId because it looks quite ugly in the swagger ui and it reveals internal package information about the application to users.

If operationId must be unique now, what is the recommended way to route calls with x-swagger-router-controller, to work around this?  Changing function names is a workaround of course, but that potentially breaks a lot of related code.  And it implies every function name referenced by swagger has to be unique across a project's entire codebase, regardless of class or package, which seems very unusual.

**Comment by sjaensch** [comment]
@monospacesoftware well that's what the Swagger 2.0 spec states:

> The id MUST be unique among all operations described in the API.

The validation we implemented is actually less strict in that it only requires unique operationIds within the same tag.

Personally I'm not familiar with `x-swagger-router-controller`. It's not supported by swagger_spec_validator or the bravado libraries.
